mlir/lib/Dialect/Vector/IR/VectorOps.cpp

```diff
@@ -1705,6 +1705,46 @@ static Value foldExtractFromBroadcast(ExtractOp extractOp) {
   return extractOp.getResult();
 }
 
+/// Fold extractOp coming from ShuffleOp.
+///
+/// Example:
+///
+///   %shuffle = vector.shuffle %a, %b [0, 8, 7, 15]
+///     : vector<8xf32>, vector<8xf32>
+///   %extract = vector.extract %shuffle[3] : f32 from vector<4xf32>
+/// ->
+///   %extract = vector.extract %b[7] : f32 from vector<4xf32>
+///
+static Value foldExtractFromShuffle(ExtractOp extractOp) {
+  // TODO: Canonicalization for dynamic position not implemented yet.
+  if (extractOp.hasDynamicPosition())
+    return Value();
```

Comment:
Out of curiosity, how do we support the dynamic case?

Reply:
Yes, that's what I'm thinking, and I'm wondering if there are other approaches. This approach looks bad to me because it introduces branching in the IR. If we don't have a good plan, perhaps we should drop the `TODO` and make an explicit comment.

Reply:
The TODO is inherited from similar canonicalizations. Not sure if branches but at least quite some instructions, including more vector extracts and select ops. Definitely not a good canonical form for dynamic cases.

Reply:
I see, thanks! Yes, I also think that it is not a good canonical form. It's good that we're on the same page..